### PR TITLE
test_stats_cluster: rewrite test cases in criterion

### LINKED ISSUE #2576: convert ut test_stats_cluster to criterion
Use Criterion to rewrite this unit test.

### PR DESCRIPTION
**Pull request by mehul-m-prajapati**
resolves #2576

### CODE REVIEW

**Comment by kira-syslogng** [comment]
This user does not have permission to start the build. Can one of the admins verify this patch and start the build?
(admin: you have the next options (make sure you checked the code):
"ok to test" to accept this pull request (and further changes) for testing
"test this please" for a one time test run
do nothing -> CI won't start)

**Comment by kira-syslogng** [comment]
This user does not have permission to start the build. Can one of the admins verify this patch and start the build?
(admin: you have the next options (make sure you checked the code):
"ok to test" to accept this pull request (and further changes) for testing
"test this please" for a one time test run
do nothing -> CI won't start)

**Comment by Kokan** [comment]
@kira-syslogng ok to test

File: lib/stats/tests/test_stats_cluster.c [hunk 1]
```diff
@@ -231,42 +223,22 @@ test_get_component_name_translates_component_to_name_properly(void)
...
+  cr_assert(stats_cluster_get_counter(sc, SC_TYPE_PROCESSED) == NULL, "get counter before tracked");
   processed = stats_cluster_track_counter(sc, SC_TYPE_PROCESSED);
-  assert_true(stats_cluster_get_counter(sc, SC_TYPE_PROCESSED) == processed, "get counter after tracked");
+  cr_assert(stats_cluster_get_counter(sc, SC_TYPE_PROCESSED) == processed, "get counter after tracked");
 
   StatsCounterItem *saved_processed = processed;
   stats_cluster_untrack_counter(sc, SC_TYPE_PROCESSED, &processed);
-  assert_true(processed == NULL, "untrack counter");
-  assert_true(stats_cluster_get_counter(sc, SC_TYPE_PROCESSED) == saved_processed, "get counter after untracked");
+  cr_assert(processed == NULL, "untrack counter");
```
Comment (Kokan) [inline lib/stats/tests/test_stats_cluster.c]: You should user `cr_assert_null`.

File: lib/stats/tests/test_stats_cluster.c [hunk 2]
```diff
@@ -141,8 +136,8 @@ _validate_yielded_counters(StatsCluster *sc, gint type, StatsCounterItem *counte
   gint t;
 
   t = va_arg(st->types, gint);
-  assert_true(t >= 0, "foreach counter returned a new counter, but we expected the end already");
-  assert_gint(type, t, "Counter type mismatch");
+  cr_assert(t >= 0, "foreach counter returned a new counter, but we expected the end already");
```
Comment (Kokan) [inline lib/stats/tests/test_stats_cluster.c]: You could use `cr_assert_geq`.

File: lib/stats/tests/test_stats_cluster.c [hunk 3]
```diff
@@ -231,42 +223,22 @@ test_get_component_name_translates_component_to_name_properly(void)
...
   StatsClusterKey sc_key;
   stats_cluster_logpipe_key_set(&sc_key, SCS_SOURCE | SCS_FILE, "id", "instance" );
   StatsCluster *sc = stats_cluster_new(&sc_key);
   StatsCounterItem *processed;
 
-  assert_true(stats_cluster_get_counter(sc, SC_TYPE_PROCESSED) == NULL, "get counter before tracked");
+  cr_assert(stats_cluster_get_counter(sc, SC_TYPE_PROCESSED) == NULL, "get counter before tracked");
   processed = stats_cluster_track_counter(sc, SC_TYPE_PROCESSED);
-  assert_true(stats_cluster_get_counter(sc, SC_TYPE_PROCESSED) == processed, "get counter after tracked");
+  cr_assert(stats_cluster_get_counter(sc, SC_TYPE_PROCESSED) == processed, "get counter after tracked");
```
Comment (Kokan) [inline lib/stats/tests/test_stats_cluster.c]: Please use `cr_assert_eq`.

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: lib/stats/tests/test_stats_cluster.c (hunk 2 above)
Comment (mehul-m-prajapati) [reply lib/stats/tests/test_stats_cluster.c]: ok

File: lib/stats/tests/test_stats_cluster.c (hunk 1 above)
Comment (mehul-m-prajapati) [reply lib/stats/tests/test_stats_cluster.c]: Yes.

File: lib/stats/tests/test_stats_cluster.c (hunk 3 above)
Comment (mehul-m-prajapati) [reply lib/stats/tests/test_stats_cluster.c]: Okay...

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: lib/stats/tests/test_stats_cluster.c [hunk 4]
```diff
@@ -231,42 +223,22 @@ test_get_component_name_translates_component_to_name_properly(void)
...
-  STATS_CLUSTER_TESTCASE(test_get_counter);
-}
-
-int
-main(int argc, char *argv[])
-{
-  test_stats_cluster();
-  return 0;
-}
+TestSuite(stats_cluster, .init = app_startup, .fini = app_shutdown);
```
Comment (MrAnno) [inline lib/stats/tests/test_stats_cluster.c]: Please note that `app_startup()` and `app_shutdown()` are not necessary here.
`app_startup()` initializes a lot of global things, so it actually transforms a nice and clean unit test into not-so-nice one.

https://github.com/balabit/syslog-ng/blob/10aabaf19a1dda43772c79b1243dec565daca6f7/lib/apphook.c#L152-L181

Comment (mehul-m-prajapati) [reply lib/stats/tests/test_stats_cluster.c]: ok. I will remove it.

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: lib/stats/tests/test_stats_cluster.c (hunk 4 above)
Comment (MrAnno) [reply lib/stats/tests/test_stats_cluster.c]: Thanks!

File: lib/stats/tests/test_stats_cluster.c [hunk 5]
```diff
@@ -231,42 +222,22 @@ test_get_component_name_translates_component_to_name_properly(void)
...
-  STATS_CLUSTER_TESTCASE(test_get_counter);
-}
-
-int
-main(int argc, char *argv[])
-{
-  test_stats_cluster();
-  return 0;
-}
+TestSuite(stats_cluster);
```
Comment (Kokan) [inline lib/stats/tests/test_stats_cluster.c]: Btw `TestSuit` is not required to write for Criterion.

Comment (mehul-m-prajapati) [reply lib/stats/tests/test_stats_cluster.c]: okay. I will keep that in mind.

**Pull request merged**